services/src/main/java/org/keycloak/storage/UserStorageManager.java

```diff
@@ -608,14 +600,13 @@ public UserModel getServiceAccount(ClientModel client) {
     }
 
     @Override
-    public Set<FederatedIdentityModel> getFederatedIdentities(UserModel user, RealmModel realm) {
+    public Stream<FederatedIdentityModel> getFederatedIdentitiesStream(UserModel user, RealmModel realm) {
         if (user == null) throw new IllegalStateException("Federated user no longer valid");
-        Set<FederatedIdentityModel> set = new HashSet<>();
-        if (StorageId.isLocalStorage(user)) {
-            set.addAll(localStorage().getFederatedIdentities(user, realm));
-        }
-        if (getFederatedStorage() != null) set.addAll(getFederatedStorage().getFederatedIdentities(user.getId(), realm));
-        return set;
+        Stream<FederatedIdentityModel> stream = StorageId.isLocalStorage(user) ?
+                localStorage().getFederatedIdentitiesStream(user, realm) : Stream.empty();
+        if (getFederatedStorage() != null)
+            stream = Streams.concat(stream, getFederatedStorage().getFederatedIdentitiesStream(user.getId(), realm));
```

Comment:
I'm wondering if we should time bound the results from federated storage same as we did in `ClientStorageManager.searchClientsByClientIdStream` whenever we concat local and federated entities.

Reply:
I will check this

Reply:
@mhajas FYI

Reply:
Well, this is a good question. However, I don't think we should. The federatedStorage is something little bit confusing, I spent quite a lot of time on investigating what is it for. If I understand correctly, this is the point: if you have some users, from external storage/provider and you are not able to store some keycloak specific properties (attributes, roles, groups etc.) in that external storage, you can configure keycloak to store it for you using this federatedStorage. For example you store a user with id `id_pointing_to_external_storage` have roles `admin` and `manager`. So whenever there is a call to federatedStorage for `FederatedIdentityModel`, it should be a call to a database/map storage (in future). See current [jpa implementation](https://github.com/keycloak/keycloak/blob/master/model/jpa/src/main/java/org/keycloak/storage/jpa/JpaUserFederatedStorageProvider.java).

Therefore, If we would time bound federated storage, the same logic would apply also to `localStorage()` and I don't like this idea. WDYT?

Reply:
Given @mhajas comment, I'm leaving it as is for now.